File: examples/person.pl
use strict;
use warnings;
use FFI::Platypus 2.00;

my $ffi = FFI::Platypus->new(
  api => 2,
  lib => './person.so',
);

$ffi->type( 'opaque' => 'person_t' );

$ffi->attach( person_new =>  ['string','unsigned int'] => 'person_t'       );
$ffi->attach( person_name => ['person_t']              => 'string'       );
$ffi->attach( person_age =>  ['person_t']              => 'unsigned int' );
$ffi->attach( person_free => ['person_t']                                  );

my $person = person_new( 'Roger Frooble Bits', 35 );

print "name = ", person_name($person), "\n";
print "age  = ", person_age($person),  "\n";

person_free($person);

File: examples/person.c
#include <string.h>
#include <stdlib.h>

typedef struct person_t {
  char *name;
  unsigned int age;
} person_t;

person_t *
person_new(const char *name, unsigned int age) {
  person_t *self = malloc(sizeof(person_t));
  self->name = strdup(name);
  self->age  = age;
}

const char *
person_name(person_t *self) {
  return self->name;
}

unsigned int
person_age(person_t *self) {
  return self->age;
}

void
person_free(person_t *self) {
  free(self->name);
  free(self);
}
